gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+		destroy_neigh_hash_table(&gk_conf->neigh_fib_front->u.neigh);
+		initialize_fib_entry(gk_conf->neigh_fib_front);
+		gk_conf->neigh_fib_front = NULL;
+	}
+	if (gk_conf->neigh6_fib_front != NULL) {
+		struct gatekeeper_if *iface = &gk_conf->net->front;
+		rte_lpm6_delete(gk_conf->lpm_tbl.lpm6,
+			iface->ip6_addr.s6_addr, iface->ip6_addr_plen);
+		destroy_neigh_hash_table(&gk_conf->neigh6_fib_front->u.neigh6);
+		initialize_fib_entry(gk_conf->neigh6_fib_front);
+		gk_conf->neigh6_fib_front = NULL;
+	}
+
+out:
+	return ret;
+}
+
+int
+setup_gk_lpm(struct gk_config *gk_conf, unsigned int socket_id)
+{
+	int ret;
+	struct rte_lpm_config ipv4_lpm_config;
+	struct rte_lpm6_config ipv6_lpm_config;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (ipv4_configured(gk_conf->net)) {
+		ipv4_lpm_config.max_rules = gk_conf->max_num_ipv4_rules;
+		ipv4_lpm_config.number_tbl8s = gk_conf->num_ipv4_tbl8s;
+
+		/*
+		 * The GK blocks only need to create one single
+		 * IPv4 LPM table on the @socket_id, so the
+		 * @lcore and @identifier are set to 0.
+		 */
+		ltbl->lpm = init_ipv4_lpm(
+			"gk", &ipv4_lpm_config, socket_id, 0, 0);
+		if (ltbl->lpm == NULL) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to initialize the IPv4 LPM table at %s\n",
+				__func__);
+			ret = -1;
+			goto out;
+		}
+
+		ltbl->fib_tbl = rte_calloc(NULL,
+			gk_conf->gk_max_num_ipv4_fib_entries,
+			sizeof(struct gk_fib), 0);
+		if (ltbl->fib_tbl == NULL) {
+			RTE_LOG(ERR, MALLOC,
+				"gk: failed to allocate the IPv4 FIB table at %s\n",
+				__func__);
+			ret = -1;
+			goto free_lpm;
+		}
+	}
+
+	if (ipv6_configured(gk_conf->net)) {
+		ipv6_lpm_config.max_rules = gk_conf->max_num_ipv6_rules;
+		ipv6_lpm_config.number_tbl8s = gk_conf->num_ipv6_tbl8s;
+
+		/*
+		 * The GK blocks only need to create one single
+		 * IPv6 LPM table on the @socket_id, so the
+		 * @lcore and @identifier are set to 0.
+		 */
+		ltbl->lpm6 = init_ipv6_lpm(
+			"gk", &ipv6_lpm_config, socket_id, 0, 0);
+		if (ltbl->lpm6 == NULL) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to initialize the IPv6 LPM table at %s\n",
+				__func__);
+			ret = -1;
+			goto free_lpm_tbl;
+		}
+
+		ltbl->fib_tbl6 = rte_calloc(NULL,
+			gk_conf->gk_max_num_ipv6_fib_entries,
+			sizeof(struct gk_fib), 0);
+		if (ltbl->fib_tbl6 == NULL) {
+			RTE_LOG(ERR, MALLOC,
+				"gk: failed to allocate the IPv6 FIB table at %s\n",
+				__func__);
+			ret = -1;
+			goto free_lpm6;
+		}
+	}
+
+	ret = init_fib_tbl(gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the FIB table at %s\n",
+			__func__);
+		goto free_lpm_tbl6;
+	}
+
+	ret = 0;
+	goto out;
+
+free_lpm_tbl6:
+	if (ipv6_configured(gk_conf->net)) {
+		rte_free(ltbl->fib_tbl6);
+		ltbl->fib_tbl6 = NULL;
+	}
+
+free_lpm6:
+	if (ipv6_configured(gk_conf->net)) {
+		destroy_ipv6_lpm(ltbl->lpm6);
+		ltbl->lpm6 = NULL;
+	}
+
+free_lpm_tbl:
+	if (ipv4_configured(gk_conf->net)) {
+		rte_free(ltbl->fib_tbl);
+		ltbl->fib_tbl = NULL;
+	}
```

Comment:
As pointed out in `setup_net_prefix_fib()`, you can change label `free_lpm_tbl` as follow:
```
free_lpm_tbl:
	if (!ipv4_configured(gk_conf->net))
		goto out;
	rte_free(ltbl->fib_tbl);
	ltbl->fib_tbl = NULL;
```

This way you can remove the test `if (ipv4_configured(gk_conf->net))` in the following label.

An analogous change goes for the IPv6 case. Of course, in the IPv6 case instead of the `goto out`, the label `free_lpm6` should point to label `free_lpm_tbl`.